File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size: int):
    data = b""
    while len(data) < size:
        received_bytes = socket.recv(size - len(data))
        if not received_bytes: raise ConnectionError("Socket closed before receiving all data")
        data += received_bytes
    return data


def send_all(socket: socket.socket, bytes: bytes):
    total_bytes_sent = 0
    while total_bytes_sent < len(bytes):
        sent_bytes = socket.send(bytes[total_bytes_sent:])
        total_bytes_sent += sent_bytes
    return total_bytes_sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"errors"
	"io"
)

func SendAll(socket io.Writer, data []byte) error {
	totalBytesSent := 0

	for totalBytesSent < len(data) {
		sentBytes, err := socket.Write(data[totalBytesSent:])
		if err != nil {
			return err
		}

		totalBytesSent += sentBytes
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	data := []byte{}

	for len(data) < size {
		buff := make([]byte, size-len(data))

		receivedBytes, err := socket.Read(buff)
		if err != nil {
			return nil, err
		}

		if receivedBytes == 0 {
			return nil, errors.New("socket closed before receiving all data")
		}

		data = append(data, buff[:receivedBytes]...)
	}

	return data, nil
}
